js/index.js

```diff
@@ -2,128 +2,798 @@
+      if (window.instrumented) {
+         /* It's important to clear it when the mouse moves from one span to another with the same parent,
+           * to clear out the parent being highlighted. */
+        clearHighlights();
+
+        /*Will also automatically highlight all nested children.*/
+        highlight($(this)); 
+
+        /*So that the parent won't be highlighted (because we are using mouseover and not mouseenter)*/
+        event.stopPropagation(); 
+
+        var path = getPath($(this));
+        var position = findPositionOf(path);
+        highlight(position); 
+      } else 
+        console.log("No instrumented type found");
+    }).on('mouseleave', 'span', function(event) {
+      clearHighlights();
+    });
+
+    $('#output').on('paste keyup', function(event) {
+      clearTimeout(typingTimer);
+      typingTimer = setTimeout(function() {
+        if(updateContent(outputElement)) {
+          eventObj.trigger('output-changed');
+        };
+      }, doneTypingInterval);
+    }).on('mouseover', 'div', function(event) {
+      if (window.reverseIndexMap) {
+        clearHighlights();
+        event.stopPropagation();
+        
+        var path = getPath(this);
+
+        //The .find() and .children() methods are similar, 
+        //except that the latter only travels a single level down the DOM tree.
+        var inputCandidates = $(inputElement).find('.' + path.join(' .'));
+
+        // Go through all input candidates and make sure the `full path` matches 
+        // (and in the same order) and then find its position to be highlighted.
+
+        $.each(inputCandidates, (function(idx, e) { 
+          var cs = getPath(e);
+          if (utils.arraysEqual(cs, path)) {
+            highlight($(e)); // highlight input
+            var p = getPath(e); // find path
+            var position = findPositionOf(p); // find the indexes in the output
+            highlight(position); // highlight them in output
+          }
+        }));
+      }
+    }).on('mouseleave', 'div', function (event) { 
+      clearHighlights(); 
+    }).on('keydown', function(event) {
+      clearTimeout(typingTimer);
+    });
+
+    $('#random').click(function () {   
+      clearTimeout(typingTimer);
+      typingTimer = setTimeout(function() {
+        generateRandom();
+      }, doneTypingInterval);
+      return false;
+    });
+
+    $("#reset").click(function() {
+      eventObj.trigger('update-url', {'schema' : '' , 'record' : ''});
+      eventObj.trigger('reset-layout');
+      window.type = undefined;
+      $('#random').addClass('-disabled-');
+      return false;
+    });
+
+    $('#uploadLink').click(function(e) {
+      console.log("uploading");
+      e.preventDefault();
+      $("#upload").trigger('click');
+      return false;
+    });
+
+    $('#upload').on("change", function(e) {
+      console.log("new file uploading");
+      var files = $('#upload')[0].files;
+      if(!!files && files.length > 0) {
```

Comment:
`!!` is redundant here, since it is only used as a predicate. (Same in a few other places.)